raft/server.cc

```diff
@@ -465,20 +468,39 @@ future<> server_impl::wait_for_entry(entry_id eid, wait_type type, seastar::abor
 
 future<entry_id> server_impl::add_entry_on_leader(command cmd, seastar::abort_source* as) {
     // Wait for sufficient memory to become available
-    const auto memory_usage = log::memory_usage_of(cmd);
-    try {
-        co_await _fsm->consume_memory(as, memory_usage);
-    } catch (semaphore_aborted&) {
-        throw request_aborted();
+    const auto requested_memory = log::memory_usage_of(cmd);
+    auto& sm = _fsm->get_log_limiter_semaphore();
+    while (true) {
+        if (sm.try_wait(requested_memory)) {
+            break;
+        }
+
+        if (_applied_idx == _fsm->log_last_idx()) {
+            logger.warn("{} max_log_size deadlock resulted in a redundant snapshot, "
+                        "you should consider increasing max_log_size (current: {}), "
+                        "decreasing snapshot_threshold (current: {}) "
+                        "or decreasing max_command_size (current: {})",
+                        _id, _config.max_log_size, _config.snapshot_threshold, _config.max_command_size);
+            co_await take_snapshot();
```

Comment:
I don't think this is the right approach, because it means we'll call
```
co_await _state_machine->take_snapshot();
```
concurrently with
```
co_await _state_machine->apply(std::move(commands));
```
or even concurrently with another `take_snapshot`.

Reply:
Look:
```
    snp.term = *_fsm->log_term_for(_applied_idx);
    snp.idx = _applied_idx;
    snp.config = _fsm->log_last_conf_for(_applied_idx);
    logger.trace("[{}] taking snapshot term={}, idx={}", _id, snp.term, snp.idx);
    snp.id = co_await _state_machine->take_snapshot();
```

what will be the index and term of the snapshot if there are concurrent entries being applied?
The last applied entry when we started calling `take_snapshot` might be `snp.idx`, but inside `take_snapshot` we may yield at the beginning and a new entry gets applied. And `snp.idx` may no longer be correct.

The entire piece assumes everything is happening sequentially.

Reply:
>  it means we'll call
> 
> co_await _state_machine->take_snapshot();
> concurrently with
> 
> co_await _state_machine->apply(std::move(commands));
> or even concurrently with another take_snapshot.

Is this a problem? I see the point in making external dependencies (`persistence`) easy to implement, the less concurrency the better, but I don't see that we promised that. I looked in the comments next to the `persistence` declaration.

> And snp.idx may no longer be correct.

Why is it incorrect? I though the only problem is that some commands might be double-applied when the server is restarted, and this is fine, isn't it?

In general, I agree that it's better to refrain from concurrency here, will push something onto `applier_fiber` queue.

Reply:
> Is this a problem? I see the point in making external dependencies (persistence) easy to implement, the less concurrency the better, but I don't see that we promised that. I looked in the comments next to the persistence declaration.

You're right, there is no promise in the comments. But already our group 0 implementation depends on the sequentiality IIUC (at least I didn't bother to think about possible concurrency when I wrote code in there). So I think we should write and maintain it, it doesn't seem that maintaining it it too much trouble for us.

> Why is it incorrect? 

Well, if a snapshot says that it was taken at index `idx`, then it should cover commands exactly up to index `idx` and no more. Otherwise the `idx` has no meaning at all. Implementations may depend on it.

> I though the only problem is that some commands might be double-applied when the server is restarted, and this is fine, isn't it?

Commands are applied starting from the last snapshot index + 1 to the last committed one. So if a particular implementation restores state machine state exactly at the last snapshot index, there will be no state onto which commands were doubly applied. And this is possible only if our snapshot indexes are correct.

Reply:
> will push something onto applier_fiber queue.

Hmm, this may become problematic; AFAIK it's a single-producer queue and it's illegal for two calls to `push_eventually` to execute concurrently.

@gleb-cloudius WDYT?

Reply:
The comment says:

```
    /// A producer-side operation. Cannot be called concurrently with other producer-side operations.
    future<> push_eventually(T&& data) noexcept;
```

Reply:
I was more curious about you suggesting some solution :P

Reply:
> I was more curious about you suggesting some solution :P

This is more complicated! May be we can signal io_fiber and push from there.

Reply:
Made a serializing wrapper on top of the queue. Maybe it's an overkill,  but I don't really like the idea of pulling a bunch of flags through fsm/io_fiber